File: leetcodes/230. Kth Smallest Element in a BST/kthSmallest.java
import java.util.ArrayList;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//dfs- inorder recursive: time O(n), space O(n)
class Solution {
  public int kthSmallest(TreeNode root, int k) {
    ArrayList<Integer> arr = inorder(root, new ArrayList<Integer>());
    return arr.get(k - 1);
  }
  private ArrayList<Integer> inorder(TreeNode root, ArrayList<Integer> arr){
    if (root == null) return arr;
    inorder(root.left, arr);
    arr.add(root.val);
    inorder(root.right, arr);
    return arr;
  }
}

//iterative inorder timeO(H+k), space O(H)
class Solution {
  public int kthSmallest(TreeNode root, int k) {
    LinkedList<TreeNode> stack = new LinkedList<>();
    while(true) {
      while(root!= null){
        stack.push(root);
        root = root.left;
      }
      root = stack.pop();
      if (--k == 0) return root.val;
      root = root.right;
    }
  }
}

File: leetcodes/230. Kth Smallest Element in a BST/kthSmallest.js
/**
 * Definition for a binary tree node.
**/
  function TreeNode(val) {
      this.val = val;
      this.left = this.right = null;
  }

/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
//bfs- inorder : time O(H+k)H: height, balanced tree O(logN + k), unbalanced tree O(N + k), spaceO(H), worse O(n), balanced tree O(logn)
 var kthSmallest = function(root, k) {
  let stack = []
  while (k) {
    while(root) {
      stack.push(root)
      root = root.left
    }
    root = stack.pop()
    k -= 1
    if (!k) {
      return root.val
    }
    root = root.right
  }
 }
//dfs - inorder timeO(n) -> collect all the n inorderly , space O(n)
var kthSmallest = function(root, k) {
  const inorder = (root, arr) => {
    //if null return arr
    if (!root) return arr
    //left
    inorder(root.left, arr)
    arr.push(root.val)
    //right
    inorder(root.right, arr)
    return arr
  }
  let result = inorder(root, [])
  return result[k - 1]
}